src/ed25519.rs

```diff
@@ -593,6 +593,91 @@ impl ExpandedSecretKey {
+    ///   which may be used to provide additional domain separation.  If not
+    ///   set, this will default to an empty string.
+    ///
+    /// # Returns
+    ///
+    /// An Ed25519ph [`Signature`] on the `prehashed_message`.
+    ///
+    /// [rfc8032]: https://tools.ietf.org/html/rfc8032#section-5.1
+    pub fn sign_prehashed<D>(&self,
+                             prehashed_message: D,
+                             public_key: &PublicKey,
+                             context: Option<&'static [u8]>) -> Signature
+        where D: Digest<OutputSize = U64> + Default
+    {
+        let mut h: D = D::default();
+        let mut hash: [u8; 64] = [0u8; 64];
+        let mut prehash: [u8; 64] = [0u8; 64];
+        let mesg_digest: Scalar;
+        let hram_digest: Scalar;
+        let r: CompressedEdwardsY;
+        let s: Scalar;
+
+        let ctx: &[u8] = match context {
+            Some(x) => x,
+            None    => b"",  // By default, the context is an empty string.
+        };
+        debug_assert!(ctx.len() <= 255, "The context must not be longer than 255 octets.");
+
+        let ctx_len: u8 = ctx.len() as u8;
+
+        // Get the result of the pre-hashed message.
+        prehash.copy_from_slice(prehashed_message.fixed_result().as_slice());
+
+        // This is the dumbest, ten-years-late, non-admission of fucking up the
+        // domain separation I have ever seen.  Why am I still required to put
+        // the upper half "prefix" of the hashed "secret key" in here?  Why
+        // can't the user just supply their own nonce and decide for themselves
+        // whether or not they want a deterministic signature scheme?  Why does
+        // the message go into what's ostensibly the signature domain separation
+        // hash?  Why wasn't there always a way to provide a context string?
+        //
+        // ...
+        //
+        // This is a really fucking stupid bandaid, and the damned scheme is
+        // still bleeding from malleability, for fuck's sake.
+        h.input(b"SigEd25519 no Ed25519 collisions");
+        h.input(&[1]); // Ed25519ph
+        h.input(&[ctx_len]);
+        h.input(ctx);
+        h.input(&self.nonce);
+        h.input(&prehash);
+        hash.copy_from_slice(h.fixed_result().as_slice());
+
+        mesg_digest = Scalar::from_bytes_mod_order_wide(&hash);
+
+        r = (&mesg_digest * &constants::ED25519_BASEPOINT_TABLE).compress();
+
+        h = D::default();
+        h.input(b"SigEd25519 no Ed25519 collisions");
+        h.input(&[1]); // Ed25519ph
+        h.input(&[ctx_len]);
+        h.input(ctx);
+        h.input(r.as_bytes());
+        h.input(public_key.as_bytes());
+        h.input(&prehash);
+        hash.copy_from_slice(h.fixed_result().as_slice());
+
+        hram_digest = Scalar::from_bytes_mod_order_wide(&hash);
+
+        s = &(&hram_digest * &self.key) + &mesg_digest;
+
+        Signature{ r: r, s: s }
```

Comment:
`Signature{ r, s }`

Reply:
you kids n ur new fangled rust syntax, get offa my lawn